### PR TITLE
Implementing copy method for container classes

### LINKED ISSUE #726: Implement a copy function for `SystemParticipants`, `RawGridElements`, `JointGridContainer` ..
For our SystemParticipants we have a copy function that enables easy copying of an object while replacing a (or multiple) specific attribute(s). It would be nice to have the functionality also implemented for our containers.

### PR DESCRIPTION
**Pull request by staudtMarius**
Resolves #726

### CODE REVIEW

File: src/main/java/edu/ie3/datamodel/models/input/container/InputContainer.java [hunk 1]
```diff
@@ -14,4 +14,44 @@
...
+
+  /**
+   * Abstract class for all builder that build child containers of interface {@link
+   * edu.ie3.datamodel.models.input.container.InputContainer}
+   *
+   * @version 3.1
+   * @since 14.02.23
+   */
+  abstract class InputContainerCopyBuilder<R extends InputEntity, E extends InputContainer<R>> {
+    protected List<R> entities;
```
Comment (t-ober) [inline src/main/java/edu/ie3/datamodel/models/input/container/InputContainer.java:29]: das Attribut wird nie genutzt und nie überschrieben, kann also weg

File: src/main/java/edu/ie3/datamodel/models/input/container/InputContainer.java [hunk 2]
```diff
@@ -14,4 +14,44 @@
...
+    /**
+     * Method to alter the list of entities directly.
+     *
+     * @param entities altered list of {@link InputEntity}'s
+     * @return child instance of {@link InputContainerCopyBuilder}
+     */
+    public InputContainerCopyBuilder<R, E> entities(List<R> entities) {
+      this.entities = entities;
+      return childInstance();
+    }
```
Comment (t-ober) [inline src/main/java/edu/ie3/datamodel/models/input/container/InputContainer.java:49]: die Methode wird nie benutzt, kann also weg

File: src/main/java/edu/ie3/datamodel/models/input/container/JointGridContainer.java [hunk 3]
```diff
@@ -76,4 +76,59 @@ public int hashCode() {
...
+     * Method to alter the {@link SubGridTopologyGraph}
+     *
+     * @param subGridTopologyGraph altered subGridTopologyGraph
+     * @return child instance of {@link JointGridContainerCopyBuilder}
+     */
+    public JointGridContainerCopyBuilder subGridTopologyGraph(
+        SubGridTopologyGraph subGridTopologyGraph) {
+      this.subGridTopologyGraph = subGridTopologyGraph;
+      return childInstance();
+    }
```
Comment (t-ober) [inline src/main/java/edu/ie3/datamodel/models/input/container/JointGridContainer.java:117]: We don't want to manually build or alter the graph. This one should be built by the PSDM method. So we can get rid of this method

File: src/main/java/edu/ie3/datamodel/models/input/container/SystemParticipants.java [hunk 4]
```diff
@@ -270,4 +274,189 @@ public int hashCode() {
...
         wecPlants);
   }
+
+  /**
+   * A builder pattern based approach to create copies of {@link SystemParticipants} containers with
+   * altered field values. For detailed field descriptions refer to java docs of {@link
+   * SystemParticipants}
+   *
+   * @version 3.1
+   * @since 14.02.23
```
Comment (t-ober) [inline src/main/java/edu/ie3/datamodel/models/input/container/SystemParticipants.java:284]: Why do we need this?

File: src/main/java/edu/ie3/datamodel/models/input/container/ThermalGrid.java [hunk 5]
```diff
@@ -46,4 +51,73 @@ public String toString() {
         + storages.size()
         + '}';
   }
+
+  /**
+   * A builder pattern based approach to create copies of {@link ThermalGrid} containers with
+   * altered field values. For detailed field descriptions refer to java docs of {@link ThermalGrid}
+   *
+   * @version 3.1
+   * @since 14.02.23
```
Comment (t-ober) [inline src/main/java/edu/ie3/datamodel/models/input/container/ThermalGrid.java:60]: s.o.

**Review by t-ober — CHANGES_REQUESTED** [review]
Nice Job!! Looks very good already. Just some minor details 🙂

File: src/main/java/edu/ie3/datamodel/models/input/container/SystemParticipants.java (hunk 4 above)
Comment (staudtMarius) [reply src/main/java/edu/ie3/datamodel/models/input/container/SystemParticipants.java:284]: The copy builder for entities, i.e ``UniqueEntitiyCopyBuilder`` also have this info in their class description. So I added them to the container copy builders.

**Comment by staudtMarius** [comment]
!test

**Review by t-ober — APPROVED** [review]
Nice work - Thanks!

**Pull request merged**